Fake Quantization support for f16 and f32

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52612 Fake Quantization support for f16 and f32**

Used the float type macro to generalize the fake_quantization per tensor functions to f16 and f64.

Summary:

Test Plan:
added test to show it works in AMP and extended the forward and backward tests below to test float16 and float64 operations. Note: the reference function doesn't work with with these types so I had to convert in and back out of these types to compare.

```
>test python test/test_quantization.py 
TestFakeQuantize.test_forward_backward_per_tensor_with_amp
...

>test python test/test_quantization.py 
TestFakeQuantize.test_forward_per_tensor_half_precision_numerics
...

>test python test/test_quantization.py TestFakeQuantize.test_forward_per_tensor_cachemask_cpu
...

>test python test/test_quantization.py TestFakeQuantize.test_backwards_per_tensor_cachemask_cpu
...

>test python test/test_quantization.py TestFakeQuantize.test_forward_per_tensor_cachemask_cuda
...

>test python test/test_quantization.py TestFakeQuantize.test_backwards_per_tensor_cachemask_cuda
...

>test python test/test_quantization.py
Ran 484 tests in 356.999s

OK (skipped=8)
```

Reviewers:

Subscribers:

Tasks: T84497471

Tags:

Differential Revision: [D26586416](https://our.internmc.facebook.com/intern/diff/D26586416)

HDCharles committed:
Fake Quantization support for f16 and f32

Summary:

Test Plan:
test python test/test_quantization.py TestFakeQuantize.test_forward_backward_per_tensor_cachemask_with_amp
test python test/test_quantization.py TestFakeQuantize.test_forward_per_tensor_cachemask_cpu
test python test/test_quantization.py TestFakeQuantize.test_backwards_per_tensor_cachemask_cpu
test python test/test_quantization.py TestFakeQuantize.test_forward_per_tensor_cachemask_cuda
test python test/test_quantization.py TestFakeQuantize.test_backwards_per_tensor_cachemask_cuda
test python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Fake Quantization support for f16 and f32"


Summary:

Test Plan:
test python test/test_quantization.py TestFakeQuantize.test_forward_backward_per_tensor_cachemask_with_amp
test python test/test_quantization.py TestFakeQuantize.test_forward_per_tensor_cachemask_cpu
test python test/test_quantization.py TestFakeQuantize.test_backwards_per_tensor_cachemask_cpu
test python test/test_quantization.py TestFakeQuantize.test_forward_per_tensor_cachemask_cuda
test python test/test_quantization.py TestFakeQuantize.test_backwards_per_tensor_cachemask_cuda
test python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on test/quantization/test_workflow_module.py:
```diff
@@ -862,22 +862,36 @@ def test_backward_per_tensor(self, device, X):
-                Y_ref = _fake_quantize_per_tensor_affine_reference(
-                    X.cpu(), scale, zero_point, quant_min, quant_max).to(device)
-                self.assertTrue(torch.allclose(Y_test, Y_ref, rtol=tolerance, atol=tolerance))
+    def _test_forward_per_tensor_cachemask_impl(self, device):
+        for float_type in (torch.float32, torch.float16, torch.float64):
+            for torch_type in (torch.qint8, torch.quint8):
+                Xs = (torch.randn(4, 8, device=device), torch.randn(4, 16, device=device)[:, ::2])
+                # pick the scale + zp so that some values get clipped
+                for X in Xs:
```
nit: there are a lot of identation levels here.  To help keep things readable, can we use `itertools.product`?  For example,

```
float_types = [torch.float32, torch.float16, torch.float64]
torch_types = [torch.qint8, torch.quint8]
Xs = [torch.randn(4, 8, device=device), torch.randn(4, 16, device=device)[:, ::2]]

for float_type, torch_type, X in itertools.product(float_types, torch_types, Xs):
  ...
```

vkuzo commented on test/quantization/test_workflow_module.py:
```diff
@@ -862,22 +862,36 @@ def test_backward_per_tensor(self, device, X):
-                scale, zero_point = obs.calculate_qparams()
-                scale, zero_point = float(scale), int(zero_point)
-                quant_min, quant_max = obs._calculate_qmin_qmax()
+    def test_forward_backward_per_tensor_cachemask_with_amp(self):
```
great!

Can we add just one more numerics test, where we ensure the fp16 calculation is numerically correct?  Something like
1. audit the reference function to ensure it handles fp16 correctly
2. create a hand crafted input and verify the reference function works for it correctly
3. pass the hand crafted input into the real function and verify it matches reference

some cases to test would be
1. normal looking tensors
2. overflow
3. underflow

raghuramank100 commented on test/quantization/test_workflow_module.py:
```diff
@@ -862,22 +862,36 @@ def test_backward_per_tensor(self, device, X):
+                    X = X.to(float_type)
+                    obs = torch.quantization.MinMaxObserver(torch_type)
+                    obs(X * 0.75)
+                    scale, zero_point = obs.calculate_qparams()
```
Since you are interested only in the numerics of the fake-quant operator, you do not need an observer.  You can directly test the operator with the values of scale and zero-point that you need. Ideally, you could sweep over a set of values for scale and zero-point. i.e something like
```
for scale, zero_point in itertools.product(scale_values, zero_point_values):
```

HDCharles replied:
That is from previous diffs where the observer is used to pick a  quant_min and quant_max such that it is guaranteed to clip the matrix. Its not used in the actual quantization test. My only change to this section is to add a for loop over torch type and apply it to X.

HDCharles committed:
Update on "Fake Quantization support for f16 and f32"


Used the float type macro to generalize the fake_quantization per tensor functions to f16 and f64.

Summary:

Test Plan:
added test to show it works in AMP and extended the forward and backward tests below to test float16 and float64 operations. Note: the reference function doesn't work with with these types so I had to convert in and back out of these types to compare.

```
>test python test/test_quantization.py 
TestFakeQuantize.test_forward_backward_per_tensor_cachemask_with_amp
...

>test python test/test_quantization.py TestFakeQuantize.test_forward_per_tensor_cachemask_cpu
...

>test python test/test_quantization.py TestFakeQuantize.test_backwards_per_tensor_cachemask_cpu
...

>test python test/test_quantization.py TestFakeQuantize.test_forward_per_tensor_cachemask_cuda
...

>test python test/test_quantization.py TestFakeQuantize.test_backwards_per_tensor_cachemask_cuda
...

>test python test/test_quantization.py
Ran 484 tests in 356.999s

OK (skipped=8)
```

Reviewers:

Subscribers:

Tasks: T84497471

Tags:

Differential Revision: [D26586416](https://our.internmc.facebook.com/intern/diff/D26586416)

[ghstack-poisoned]

vkuzo reviewed (approved):
lgtm, thanks for working on this
